spec.emu

```diff
@@ -47,24 +27,59 @@ contributors: Jordan Harband
         <emu-alg>
           1. Let _str_ be ? ToString(_S_).
           1. Let _cpList_ be StringToCodePoints(_str_).
-          1. Let _punctuators_ be the following String, which consists of every ASCII punctuator except U+005F (LOW LINE): *"(){}[]|,.?\*+-^$=<>\/#&!%:;@~'"`"*.
-          1. Let _toEscape_ be StringToCodePoints(_punctuators_).
           1. Let _escapedList_ be a new empty List.
           1. For each code point _c_ in _cpList_, do
             1. If _escapedList_ is empty and _c_ is matched by |DecimalDigit|, then
-              1. Append code unit U+005C (REVERSE SOLIDUS) to _escapedList_.
-              1. Append code unit U+0078 (LATIN SMALL LETTER X) to _escapedList_.
-              1. Append code unit U+0033 (DIGIT THREE) to _escapedList_.
-            1. Else if _toEscape_ contains _c_ or _c_ is matched by |WhiteSpace|, then
-              1. Append code unit U+005C (REVERSE SOLIDUS) to _escapedList_.
-            1. Append _c_ to _escapedList_.
+              1. Append code point U+005C (REVERSE SOLIDUS) to _escapedList_.
+              1. Append code point U+0078 (LATIN SMALL LETTER X) to _escapedList_.
+              1. Append code point U+0033 (DIGIT THREE) to _escapedList_.
+              1. Append _c_ to _escapedList_.
+            1. Else,
+              1. Append the code points in EncodeForRegExpEscape(_c_) to _escapedList_.
           1. Return CodePointsToString(_escapedList_).
         </emu-alg>
 
         <emu-note>
           <p>`escape` takes a string and escapes it so it can be literally represented as a pattern. In contrast EscapeRegExpPattern (as the name implies) takes a pattern and escapes it so that it can be represented as a string. While the two are related, they do not share the same character escape set or perform similar actions.</p>
         </emu-note>
       </emu-clause>
+
+      <emu-clause id="sec-encode" type="abstract operation">
+        <h1>
+          EncodeForRegExpEscape (
+            _c_: a code point,
+          ): a List of code points
+        </h1>
+        <dl class="header">
+          <dt>description</dt>
+          <dd>If the code unit represents a RegExp punctuator that needs escaping, or ASCII whitespace, it produces the code units for *"\x"* followed by the relevant escape code. If the code unit represents non-ASCII white space, it produces the code units for *"\u"* followed by the relevant escape code. Otherwise, it returns a List containing the original code unit.</dd>
```

Comment:
```suggestion
          <dd>If _c_ represents a RegExp punctuator that needs escaping or ASCII whitespace, it produces the code points for *"\x"* followed by the relevant escape code. If _c_ represents non-ASCII white space, it produces the code points for *"\u"* followed by the relevant escape code. Otherwise, it returns a singleton List containing _c_.</dd>
```

Reply:
The _c_ replacements seem good; i think the comma is necessary for it to read properly, and what is a "singleton" List vs any other kind of List?

Reply:
A singleton list has one element. This is to be explicit that it's not any old list that contains `c` (and possibly other things) but the list that contains *only* `c`.

Reply:
That's a really confusing word to use to me; in JS a singleton is when you have a class with a single shared instance. Is that qualifier necessary, given that it already says "a List containing c"?

Reply:
It's not necessary, I just thought it was a cheap way to be more precise. https://en.wikipedia.org/wiki/Singleton_(mathematics)

Reply:
I don't see how it adds precision, since the prose already explicitly says it has one item in it.

Reply:
Where does it say that?

Reply:
isn't "Otherwise, it returns a List containing the original code unit" that?

either way i'd like to land this, and i'm happy to have a PR tweaking it further, or we can do it at 262 time.

Reply:
> isn't "Otherwise, it returns a List containing the original code unit" that?

I still don't know where this prose you're referring to is.

`[a]` is a list containing `a`, and is also a singleton list containing `a`. `[a, b]` is another list containing `a`, but is not a singleton list.

Anyway, it's not important, but for me it seemed like an easy win for precision.